File: 006-redos-safety/engines/redos.go
// Go の regexp は RE2 系で、バックトラッキングをしない
package main

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
)

type row struct {
	N   int     `json:"n"`
	Sec float64 `json:"sec"`
}

func main() {
	re := regexp.MustCompile(`(a+)+$`)
	var rows []row
	for _, n := range []int{16, 20, 24, 26, 28} {
		subject := strings.Repeat("a", n) + "!"
		start := time.Now()
		re.MatchString(subject)
		sec := time.Since(start).Seconds()
		rows = append(rows, row{n, sec})
		fmt.Printf("  n=%3d  %.6f s\n", n, sec)
	}
	out, _ := json.Marshal(map[string]any{"engine": "Go regexp (RE2)", "results": rows})
	fmt.Println(string(out))
}

File: 006-redos-safety/engines/redos.mjs
// JavaScript（V8）でのバックトラッキングの伸び方を測る
const LENGTHS = [16, 20, 24, 26, 28];
const results = [];
for (const n of LENGTHS) {
  const subject = 'a'.repeat(n) + '!';
  const start = process.hrtime.bigint();
  /(a+)+$/.test(subject);
  const sec = Number(process.hrtime.bigint() - start) / 1e9;
  results.push({ n, sec });
  console.log(`  n=${String(n).padStart(3)}  ${sec.toFixed(6)} s`);
  if (sec > 10) { console.log('  （10 秒を超えたため打ち切り）'); break; }
}
console.log(JSON.stringify({ engine: `Node ${process.version} (V8)`, results }));
